napari/layers/vectors/_slice.py

```diff
@@ -0,0 +1,119 @@
+from dataclasses import dataclass, field
+from typing import Any, Union
+
+import numpy as np
+
+from napari.layers.utils._slice_input import _SliceInput
+
+
+@dataclass(frozen=True)
+class _VectorSliceResponse:
+    """Contains all the output data of slicing an Vectors layer.
+
+    Attributes
+    ----------
+    indices : array like
+        Indices of the sliced Vectors data.
+    alphas: array like or scalar
+        Used to change the opacity of the sliced vectors for visualization.
+        Should be broadcastable to indices.
+    dims : _SliceInput
+        Describes the slicing plane or bounding box in the layer's dimensions.
+    """
+
+    indices: np.ndarray = field(repr=False)
+    alphas: Union[np.ndarray, float] = field(repr=False)
+    dims: _SliceInput
+
+
+@dataclass(frozen=True)
+class _VectorSliceRequest:
+    """A callable that stores all the input data needed to slice a Vectors layer.
+
+    This should be treated a deeply immutable structure, even though some
+    fields can be modified in place. It is like a function that has captured
+    all its inputs already.
+
+    In general, the calling an instance of this may take a long time, so you may
+    want to run it off the main thread.
+
+    Attributes
+    ----------
+    dims : _SliceInput
+        Describes the slicing plane or bounding box in the layer's dimensions.
+    data : Any
+        The layer's data field, which is the main input to slicing.
+    dims_indices : tuple of ints or slices
+        The slice indices in the layer's data space.
+    others
+        See the corresponding attributes in `Layer` and `Vectors`.
+    """
+
+    dims: _SliceInput
+    data: Any = field(repr=False)
+    dims_indices: Any = field(repr=False)
+    length: float = field(repr=False)
+    out_of_slice_display: bool = field(repr=False)
+
+    def __call__(self) -> _VectorSliceResponse:
+        # Return early if no data
+        if len(self.data) == 0:
+            return _VectorSliceResponse(
+                indices=np.empty(0, dtype=int),
+                alphas=np.empty(0),
+                dims=self.dims,
+            )
+
+        not_disp = list(self.dims.not_displayed)
+        if not not_disp:
+            # If we want to display everything, then use all indices.
+            # alpha is only impacted by not displayed data, therefore 1
+            return _VectorSliceResponse(
+                indices=np.arange(len(self.data), dtype=int),
+                alphas=1,
+                dims=self.dims,
+            )
+
+        # We want a numpy array so we can use fancy indexing with the non-displayed
+        # indices, but as self.dims_indices can (and often/always does) contain slice
+        # objects, the array has dtype=object which is then very slow for the
+        # arithmetic below. As Vectors._round_index is always False, we can safely
```

Comment:
Unlike `Points`, `Vectors._round_index` is `True`, so the comment is a bit misleading here. But maybe it should be `False` given that slicing acts in a point/vertex-like way.

In fact, I'd probably argue that it's just `_ImageBase` which should have `_round_index` be `True`. But I'll leave that discussion for later!